xtra-libp2p-ping/src/protocol.rs

```diff
@@ -0,0 +1,46 @@
+//! The actual protocol functions for sending ping messages.
+//!
+//! Insbired by https://github.com/libp2p/rust-libp2p/blob/102509afe3a3b984e43a88dbe4de935fde36f319/protocols/ping/src/protocol.rs#L82-L113.
+
+pub(crate) const SIZE: usize = 32;
+
+use futures::AsyncReadExt;
+use futures::AsyncWriteExt;
+use rand::distributions;
+use rand::thread_rng;
+use rand::Rng;
+use std::io;
+use std::time::Duration;
+use std::time::Instant;
+use xtra_libp2p::Substream;
+
+/// Sends a ping and waits for the pong.
+pub(crate) async fn send(mut stream: Substream) -> io::Result<Duration> {
+    let payload: [u8; SIZE] = thread_rng().sample(distributions::Standard);
+    stream.write_all(&payload).await?;
+    stream.flush().await?;
+
+    let started = Instant::now();
+
+    let mut recv_payload = [0u8; SIZE];
+    stream.read_exact(&mut recv_payload).await?;
+
+    if recv_payload != payload {
+        return Err(io::Error::new(
+            io::ErrorKind::InvalidData,
+            "Ping payload mismatch",
+        ));
+    }
+
+    Ok(started.elapsed())
+}
+
+/// Waits for a ping and sends a pong.
+pub(crate) async fn recv(mut stream: Substream) -> io::Result<()> {
```

Comment:
❓ What is the point of separating these functions into `protocol`?

I know that the initial design also did that, but I have still not figure out what the advantage would be. I think it is way easier to understand if we inlined this code in the respective actor. 
For me the actor represents the protocol - there is not really a way (at least I don't see a clear one yet...) to separate "the protocol" (as in the pure message exchange) from the actor. With ping now using two actors it makes even less sense to me to have these functions separate from the actor. I feel it just makes it harder to follow the code.

Reply:
The ping protocol we are performing doesn't have anything to do with the substream that we perform it through.

I've introduced 18dbe20e8c12fcd09fb4743518c5c872ffd740b5 to demonstrate how this doesn't need to depend on `xtra`.

The message protocol is on a different level of abstraction than the actors that call its functions. For example, the `ping` actor calculates latencies based on performing the protocol. The message protocol could also be tested in isolation.

Reply:
It will be interesting to "map" this information to other protocols we design. For rollover we somehow have a similar constraint:
The `setup_contract::roll_over` function is the "protocol" and everything else is only scaffolding to *start* that protocol - including the maker's accept and reject. 

But, when it comes to introducing a new version of the protocol then that would be at level of negotiating the substream - and that is done by the actor. So, to have a new protocol version we would need another actor as well, which renders this separation somewhat useless. Or am I missing something?